Node repurpose tool

henningandersen opened the pull request:
When a node is repurposed to master/no-data or no-master/no-data, v7.x
will not start (see #37748 and #37347). The `elasticsearch-node repurpose`
tool can fix this by cleaning up the problematic data.

Unsure if this should be backported to 7.0 too?

Documentation to be done in a follow-up PR.

henningandersen committed:
Node repurpose tool

When a node is repurposed to master/no-data or no-master/no-data, v7.x
will not start (see #37748 and #37347). The `elasticsearch repurpose`
tool can fix this by cleaning up the problematic data.

elasticmachine commented:
Pinging @elastic/es-distributed

andrershov commented on server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java:
```diff
@@ -0,0 +1,100 @@
+import static org.mockito.Matchers.contains;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class NodeRepurposeCommandIT extends ESIntegTestCase {
```
Since it's an integration test, I think we need to be able to verify that the document that we have indexed is lost. For example, you may start one master-only node and one data-only node and then repurpose data-only to be master-only or no-master no-data. In this case, when the repurposed node joins the cluster, the index will be re-created but the document is gone.  
Also due to randomization, `executeRepurposeCommand` is a bit hard to follow. I think here we can stop a particular node using `nameFilter` and explicitly run repurpose tool successfully on this node and unsuccessfully on another node.

henningandersen replied:
Regarding checking for lost data, I added a check that we loose the index completely in f5396b18519b6d8115852673b43fe692dade310c , I think that should suffice, do you agree?

Commit 85cf0e65edcaefb2f80280dd936b7c24839e0eae fixes test to stop one node, run the tool successfully on that node and unsuccessfully on the other node.

andrershov replied:
I'm not sure how the test works, but after you restart two nodes - they should not form the cluster, because there is no longer the master node. It'd nice to figure out why ensureGreen returns successfully in this case.
But in general, I think, having one dedicated master-eligible node and one data node and repurposing of data node (mentioned in the comment) would be easier to follow.

henningandersen replied:
@andrershov this works because `InternalTestCluster` maintains the voting configuration in `InternalTestCluster.excludeMasters`, which takes the stopping node of the voting configuration. This is used by several other tests (is the default for `InternalTestCluster`), for instance I think `GatewayIndexStateIT.testIndexDeletionWhenNodeRejoins` depends on the same.

henningandersen replied:
Changed to one master only and one data only node in 7a0712c9e2f6df80e83d7158ce99b27cf50c88c3.

henningandersen replied:
Sorry, the test needed more adjustments, fixed in 4cbf288dab44ebb7422437fcccefeb67790797a6

I somehow ran the wrong test 100 times ;-)

andrershov commented on server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandTests.java:
```diff
@@ -0,0 +1,325 @@
+        assertEquals("Must not touch files", before, after);
+    }
+
+    private long digestPaths() {
```
I wonder if we already have a similar function implemented somewhere (it takes 40 lines of code here) or probably it makes sense to add it to some util class. Probably @ywelsch can suggest something.

henningandersen replied:
I found none, will await feedback from @ywelsch on this.

ywelsch replied:
I'm not aware of any such existing functionality. There is `BufferedChecksumIndexInput` in Lucene, but that will not shorten the code here.

andrershov commented on server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java:
```diff
@@ -0,0 +1,100 @@
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class NodeRepurposeCommandIT extends ESIntegTestCase {
+
+    public void testRepurpose() throws Exception {
```
Since we have tests for `RepurposeCommand` not in `ElasticsearchNodeCommandIT` can we rename `ElasticsearchNodeCommandIT` to make it clear that it tests unsafe bootstrap and detach cluster?

henningandersen replied:
Fixed in 4df583c2cf7ed901b0ee6f20e156c7e70f85e855

andrershov commented on server/src/main/java/org/elasticsearch/cluster/coordination/NodeToolCli.java:
```diff
@@ -36,6 +37,7 @@ public NodeToolCli() {
+        subcommands.put("repurpose", new NodeRepurposeCommand());
     }
 
     public static void main(String[] args) throws Exception {
```
The description that we pass to super when constructing NodeToolCli no longer looks correct.
Don't forget about `ArchiveTestCase` that tests this message

henningandersen replied:
Changed it to be more generic to cover all 3 tools:
`A CLI tool to do unsafe cluster and index manipulations on current node`

See 85cf0e65edcaefb2f80280dd936b7c24839e0eae

andrershov replied:
This description once again makes me think that detach/unsafe bootstrap and repurpose should be two different tools. However, let's leave it as it is for now.

andrershov reviewed (changes requested):
@henningandersen thanks for your PR! I've left several comments, overall looks good already.

henningandersen committed:
Node repurpose tool

Moved "how to get verbose info" message to after summary message.
Test case now randomly creates up to 10 shards.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into feature_node_repurpose_tool

henningandersen committed:
Merge remote-tracking branch 'origin/master' into feature_node_repurpose_tool

henningandersen committed:
Node repurpose tool

Renamed ElasticSearchNodeCommandIT to avoid confusion with repurpose
tool that has a separate IT.

henningandersen committed:
Node repurpose tool

Changed NodeRepurposeCommandIT to be more specific about which node it
runs the tool on rather than just try all.
Fixed NodeToolCli messaging.

henningandersen committed:
Node repurpose tool

Changed NodeRepurposeCommandIT to validate indexes still exists after
repurpose and do not exist after repurpose on all nodes.

henningandersen commented on server/src/main/java/org/elasticsearch/env/NodeRepurposeCommand.java:
```diff
@@ -0,0 +1,238 @@
+    @Override
+    protected void processNodePaths(Terminal terminal, Path[] dataPaths, Environment env) throws IOException {
+        assert DiscoveryNode.isDataNode(env.settings()) == false;
+
```
Fixed in 18df8d71aa3d1ec27090005cad512e238c3c3692

henningandersen commented on server/src/main/java/org/elasticsearch/env/NodeRepurposeCommand.java:
```diff
@@ -0,0 +1,238 @@
+        Set<String> indexUUIDs = indexUUIDsFor(indexPaths);
+        outputVerboseInformation(terminal, nodePaths, indexPaths, indexUUIDs);
+
+        terminal.println(noMasterMessage(indexUUIDs.size(), shardDataPaths.size(), indexMetaDataPaths.size()));
```
I intentionally had the longer list of indices and shards before the summary, so I kept that but moved "Use -v to see list of paths and indices affected" to be output after the summary.

See 18df8d71aa3d1ec27090005cad512e238c3c3692

henningandersen commented on server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandTests.java:
```diff
@@ -0,0 +1,325 @@
+                .numberOfReplicas(1)
+                .build(), env.indexPaths(INDEX));
+            for (Path path : env.indexPaths(INDEX)) {
+                if (includeShardData) {
```
Fixed in 18df8d71aa3d1ec27090005cad512e238c3c3692

henningandersen commented on server/src/main/java/org/elasticsearch/env/NodeRepurposeCommand.java:
```diff
@@ -0,0 +1,238 @@
+    static final String NO_CLEANUP = "Node has node.data=true -> no clean up necessary";
+    static final String NO_DATA_TO_CLEAN_UP_FOUND = "No data to clean-up found";
+    static final String NO_SHARD_DATA_TO_CLEAN_UP_FOUND = "No shard data to clean-up found";
+    static final String PRE_V7_MESSAGE =
```
Fixed in 18df8d71aa3d1ec27090005cad512e238c3c3692

henningandersen commented on server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandTests.java:
```diff
@@ -0,0 +1,325 @@
+
+    @Before
+    public void createNodePaths() throws IOException {
+        settings = buildEnvSettings(Settings.EMPTY);
```
Fixed in 18df8d71aa3d1ec27090005cad512e238c3c3692

henningandersen commented on server/src/main/java/org/elasticsearch/env/NodeRepurposeCommand.java:
```diff
@@ -0,0 +1,238 @@
+    }
+
+    private String toIndexName(NodeEnvironment.NodePath[] nodePaths, String uuid) {
+        Index index = new Index("dummy", uuid);
```
Fixed in 18df8d71aa3d1ec27090005cad512e238c3c3692

henningandersen commented:
Thanks for a thorough review, @andrershov , I believe I addressed all comments, so is ready for another round.

andrershov reviewed (changes requested):
Thanks for the changes @henningandersen. I've left a comment about the test and would like to review it again after it's addressed.

henningandersen committed:
Node repurpose tool

Changed NodeRepurposeCommandIT to be less random according to review
comments.

henningandersen commented:
> Thanks for the changes @henningandersen. I've left a comment about the test and would like to review it again after it's addressed.

Thanks @andrershov , I modified the test, please have another look.

henningandersen committed:
Node repurpose tool

Fixed NodeRepurposeCommandIT.

henningandersen commented:
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/default-distro

henningandersen committed:
Merge remote-tracking branch 'origin/master' into feature_node_repurpose_tool

henningandersen committed:
Merge remote-tracking branch 'origin/master' into feature_node_repurpose_tool

henningandersen commented:
@elasticmachine run elasticsearch-ci/packaging-sample

henningandersen committed:
Merge commit '7893aaec0c2faaf202212ad5f4ad0ee2333378d4' into feature_node_repurpose_tool

henningandersen committed:
Merge remote-tracking branch 'origin/master' into feature_node_repurpose_tool

andrershov commented on server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java:
```diff
@@ -68,29 +68,25 @@ public void testRepurpose() throws Exception {
-        executeRepurposeCommandForOrdinal(noMasterNoDataSettings, indexUUID, 0);
+        executeRepurposeCommandForOrdinal(noMasterNoDataSettings, indexUUID, 0, 0);
 
         internalCluster().startMasterOnlyNode();
```
This is not easy to follow. We've repurposed both nodes as non-master and non-data. And now we're starting them as "master" and "data" correspondingly. 
I think, that we at least need to add a comment that describes what we're doing here. 
Like "now both nodes are coordinating, let's start one of them as a master node and the second one as the data node, this does not require repurpose tool invocation. However, since they were converted to coordinating nodes previously, all data and metadata should be lost and we can verify it".

andrershov commented on server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java:
```diff
@@ -0,0 +1,112 @@
+
+        ensureGreen();
+
+        // indexes gone.
```
index is gone

andrershov reviewed (approved):
@henningandersen I've left several comments to the test, but no need for the next pass.
Since there is no feedback from @yannick on "digest" function, let's have this custom implementation for now.
As for backporting to 7.0, on the one hand, it would be nice to have it in 7.0, but I'm affraid we don't have enough time for documentation. So I think backport to 7.1 is enough.
@yannick what do you think?

henningandersen committed:
Node repurpose tool

Improvements to NodeRepurposeCommandIT.

ywelsch commented:
Given that #37748 and #37347 are in 7.0, we should try to get this in there as well. I'm sure we will figure out the docs before release.
